upup/pkg/fi/cloudup/terraform/target_hcl2.go

```diff
@@ -87,7 +84,22 @@ func (t *TerraformTarget) finish012(taskMap map[string]fi.Task) error {
 
 	terraformBlock := rootBody.AppendNewBlock("terraform", []string{})
 	terraformBody := terraformBlock.Body()
-	terraformBody.SetAttributeValue("required_version", cty.StringVal(">= 0.12.0"))
+	terraformBody.SetAttributeValue("required_version", cty.StringVal(">= 0.12.26"))
+
+	requiredProvidersBlock := terraformBody.AppendNewBlock("required_providers", []string{})
+	requiredProvidersBody := requiredProvidersBlock.Body()
+
+	if t.Cloud.ProviderID() == kops.CloudProviderGCE {
+		writeMap(requiredProvidersBody, "google", map[string]cty.Value{
+			"source":  cty.StringVal("hashicorp/google"),
+			"version": cty.StringVal(">= 3.44.0"),
+		})
+	} else if t.Cloud.ProviderID() == kops.CloudProviderAWS {
+		writeMap(requiredProvidersBody, "aws", map[string]cty.Value{
+			"source":  cty.StringVal("hashicorp/aws"),
+			"version": cty.StringVal(">= 3.12.0"),
```

Comment:
By increasing the version constraints on the aws and google providers we're forcing anyone to upgrade their provider version when they upgrade to this version of kops. Is that really necessary? Ideally we could find the oldest 3.X provider version that supports all of the functionality that kops uses and use that in the version constraint.

I know I've had issues upgrading aws provider minor versions in the past but not with any resource types used by kops.

Just from memory, I think one of the most recent changes to kops' terraform output that requires a recent provider version would be https://github.com/kubernetes/kops/pull/9533 and [2.36.0](https://github.com/terraform-providers/terraform-provider-aws/blob/release/2.x/CHANGELOG.md#2460-january-23-2020) but I don't recall if newer terraform versions require aws provider 3.X.

Reply:
Agreed. Probably best not to force too much change on developers using Kops v1.20.

After some testing the minimum provider versions to create a valid Terraform plan consisting of a simple Kops cluster is `2.46.0` for AWS and `2.19.0` for Google.

AWS Terraform provider `2.45.0` generates the following error...

```
Error: Unsupported argument

  on kubernetes.tf line 292, in resource "aws_key_pair" "kubernetes-training-kops-k8s-local-3e19983429436e7b1ff18156a91241c6":
 292:   tags = {

An argument named "tags" is not expected here.
```

Relevant scripts updated and ready for review.